subsys/net/lib/sockets/sockets_can.c

```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdbool.h>
+#include <fcntl.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_sock_can, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+#include <kernel.h>
+#include <entropy.h>
+#include <misc/util.h>
+#include <net/net_context.h>
+#include <net/net_pkt.h>
+#include <net/socket.h>
+#include <syscall_handler.h>
+#include <misc/fdtable.h>
+#include <net/socket_can.h>
+
+#include "sockets_internal.h"
+
+extern const struct socket_op_vtable sock_fd_op_vtable;
+
+static const struct socket_op_vtable can_sock_fd_op_vtable;
```

Comment:
Just following the convention here, the tls and raw socket support is done same way.

Reply:
> why couldn't this can specific stuff go directly into socket.c ?

The whole idea is to avoid patching plain socket.c on adding each new socket type. It still need to be patched to overload constructor (i.e. socket()), and we need to implement a solution for that.

Reply:
Remark: I'm currently working on ISO-TP for zephyr. Maybe you should keep in mind that RAW CAN is not the only protocol for CAN.

Reply:
> Remark: I'm currently working on ISO-TP for zephyr.

Good to know. I actually removed the ISO-TP address from the sockaddr_can struct as there was no use for it in this PR.

Reply:
@pfalcon @jukka tls in itself is so specific that it indeed makes sense to split. But raw socket and can? 95% of the code is plain same. Unless we don't give a damn about small targets...

Reply:
But this will take some time. I have to touch the stm23 CAN driver because it can happen that data is sent out of order. This is because the mailbox sending priority is handled by hardware. This probably happens to the can raw socket too if you fill up all mailboxes without waiting (can_send callback is not NULL).

Reply:
@tbursztyka 

> 95% of the code is plain same. 

I so far haven't seen this (need to review in more detail). As a pre-statement, I seriously doubt that there's much in common in recv() implementation for TCP and RAW. Maybe there's more in common of recv() implementation for UDP and RAW. We certainly should make reusable code to be actually reusable. We probably can merge handling of some different socket types into single implementation (like is currently done with TCP and UDP). All that can be premature optimization, or artifact of subideal implementation. For example, I can't imagine CAN and normal IP to have much in common re: implementation.

Anyway, again, let me have a deeper look first. And we bigger problem actually: lack of clear understanding of what's actually being implemented. (I generalize that even to authors of PRs, because they can't clearly convey that information, which I treat as a sign of not clear enough understanding, sorry).